src/util.c

```diff
@@ -132,44 +132,303 @@ long PURE str_to_long(const char *str)
-    while (!(x & 0x80000000U) && count < 34) {
-        x <<= 1;
-        count++;
+    // Compiler builtins typically return int, but we use unsigned internally
+    // to reduce the number of guards we see in the proofs.
+    unsigned count = 32;
+    uint32_t mask = UINT32_MAX;
+
+    // Each iteration i (counting backwards) considers the least significant
+    // 2^(i+1) bits of x as the current focus. At the first iteration, the
+    // focus is the whole input. Each iteration assumes x contains no 1 bits
+    // outside its focus. The iteration contains a test which determines
+    // whether there are any 1 bits in the upper half (2^i bits) of the focus,
+    // setting `bits` to 2^i if there are, or zero if not. Shifting by `bits`
+    // then narrows the focus to the lower 2^i bits and satisfies the
+    // assumption for the next iteration. After the final iteration, the focus
+    // is just the least significant bit, and the most significsnt 1 bit of the
+    // original input (if any) has been shifted into this position. The leading
+    // zero count can be determined from the total shift.
+    //
+    // The iterations are given a very regular structure to facilitate proofs,
+    // while also generating reasonably efficient binary code.
+
+    // The `if (1)` blocks make it easier to reason by chunks in the proofs.
+    if (1) {
+        // iteration 4
+        mask >>= (1 << 4); // 0x0000ffff
+        unsigned bits = ((unsigned)(mask < x)) << 4; // [0, 16]
+        x >>= bits; // <= 0x0000ffff
+        count -= bits; // [16, 32]
+    }
+    if (1) {
+        // iteration 3
+        mask >>= (1 << 3); // 0x000000ff
+        unsigned bits = ((unsigned)(mask < x)) << 3; // [0, 8]
+        x >>= bits; // <= 0x000000ff
+        count -= bits; // [8, 16, 24, 32]
     }
-    return count;
+    if (1) {
+        // iteration 2
+        mask >>= (1 << 2); // 0x0000000f
+        unsigned bits = ((unsigned)(mask < x)) << 2; // [0, 4]
+        x >>= bits; // <= 0x0000000f
+        count -= bits; // [4, 8, 12, ..., 32]
+    }
+    if (1) {
+        // iteration 1
+        mask >>= (1 << 1); // 0x00000003
+        unsigned bits = ((unsigned)(mask < x)) << 1; // [0, 2]
+        x >>= bits; // <= 0x00000003
+        count -= bits; // [2, 4, 6, ..., 32]
+    }
+    if (1) {
+        // iteration 0
+        mask >>= (1 << 0); // 0x00000001
+        unsigned bits = ((unsigned)(mask < x)) << 0; // [0, 1]
+        x >>= bits; // <= 0x00000001
+        count -= bits; // [1, 2, 3, ..., 32]
+    }
+
+    // If the original input was zero, there will have been no shifts, so this
+    // gives a result of 32. Otherwise, x is now exactly 1, so subtracting from
+    // count gives a result from [0, 1, 2, ..., 31].
+    return count - x;
 }
 
-uint32_t __ctzsi2(uint32_t x)
+static inline unsigned clz64(uint64_t x)
 {
-    uint32_t count = 0;
-    while (!(x & 0x000000001) && count <= 32) {
-        x >>= 1;
-        count++;
+    unsigned count = 64;
```

Comment:
count is "unsigned int" and probably 32-bit. but using uint32_t directly should convey the intention clearly?

Reply:
The counts are always small (64 will be the biggest count we see), so any integer size will do. My intention here was to let the compiler pick its "default" integer size, whether that's 32 or 64 bits. Then, at least in principle, compilers can be configured with whatever default size is most efficient for the architecture.